sftp/sftp.go

```diff
@@ -101,6 +96,77 @@ func (fm *fileManagerImpl) Delete(remoteFilePath string) error {
 	if err != nil {
 		return fmt.Errorf("cannot delete file: %w", err)
 	}
+	return nil
+}
 
+func (fm *fileManagerImpl) Reset() error {
+	newClient, err := newSFTPClientFromConfig(fm.config)
+	if err != nil {
+		return err
+	}
+	fm.client = newClient
 	return nil
 }
+
+func NewFileManager(config *SSHConfig) (FileManager, error) {
+	sftpClient, err := newSFTPClientFromConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return &fileManagerImpl{client: sftpClient, config: config}, nil
+}
+
+type retryableFileManagerImpl struct {
```

Comment:
I think we can have a leaner and clearer implementation. I'll try to explain, let me know your thoughts.

Instead of having two almost equivalent implementations, one called `fileManagerImpl` and one `retryableFileManagerImpl`, why don't we keep one and pass an option to the constructor called `WithRetriesOnConnectionLost`.

Benefits of the approach:
* `NewRetryableFileManager` gives you a retryable file manager but it's not really clear what it retries, since its name is vague, I would imply that it retries for everything but that is not the case.
* We keep only one implementation and check a boolean flag to know whether we should retry on connection lost.
* The `WithRetriesOnConnectionLost` option can't be mistaken. Its behaviour would be quite clear. Also it could easily take a few parameters to specify intervals and other configurations that might apply to retries.

Reply:
It would look like this more or less:

```diff
diff --git a/sftp/sftp.go b/sftp/sftp.go
index 7ba55d3..d605688 100644
--- a/sftp/sftp.go
+++ b/sftp/sftp.go
@@ -25,10 +25,19 @@ type FileManager interface {
 	Reset() error
 }
 
+type Option func(impl *fileManagerImpl)
+
+func WithRetryOnConnectionLost(retryOnConnectionLost bool) Option {
+	return func(impl *fileManagerImpl) {
+		impl.retryOnConnectionLost = retryOnConnectionLost
+	}
+}
+
 // fileManagerImpl is a real implementation of FileManager
 type fileManagerImpl struct {
-	client Client
-	config *SSHConfig
+	client                Client
+	config                *SSHConfig
+	retryOnConnectionLost bool
 }
 
 // Upload uploads a file to the remote server
@@ -92,6 +101,15 @@ func (fm *fileManagerImpl) Download(remoteFilePath, localDir string) error {
 
 // Delete deletes a file on the remote server
 func (fm *fileManagerImpl) Delete(remoteFilePath string) error {
+	if fm.retryOnConnectionLost {
+		return fm.retryOnLostConnection(func() error {
+			return fm.delete(remoteFilePath)
+		})
+	}
+	return fm.delete(remoteFilePath)
+}
+
+func (fm *fileManagerImpl) delete(remoteFilePath string) error {
 	err := fm.client.Remove(remoteFilePath)
 	if err != nil {
 		return fmt.Errorf("cannot delete file: %w", err)
@@ -108,50 +126,25 @@ func (fm *fileManagerImpl) Reset() error {
 	return nil
 }
 
-func NewFileManager(config *SSHConfig) (FileManager, error) {
+func NewFileManager(config *SSHConfig, opts ...Option) (FileManager, error) {
 	sftpClient, err := newSFTPClientFromConfig(config)
 	if err != nil {
 		return nil, err
 	}
-	return &fileManagerImpl{client: sftpClient, config: config}, nil
-}
-
-type retryableFileManagerImpl struct {
-	fileManager FileManager
-}
-
-func (r *retryableFileManagerImpl) Upload(localFilePath, remoteFilePath string) error {
-	fileOperation := func() error {
-		return r.fileManager.Upload(localFilePath, remoteFilePath)
-	}
-	return r.retryOnConnectionLost(fileOperation)
-}
-
-func (r *retryableFileManagerImpl) Download(remoteFilePath, localDir string) error {
-	fileOperation := func() error {
-		return r.fileManager.Download(remoteFilePath, localDir)
-	}
-	return r.retryOnConnectionLost(fileOperation)
-}
-
-func (r *retryableFileManagerImpl) Delete(remoteFilePath string) error {
-	fileOperation := func() error {
-		return r.fileManager.Delete(remoteFilePath)
+	fm := &fileManagerImpl{client: sftpClient, config: config}
+	for _, opt := range opts {
+		opt(fm)
 	}
-	return r.retryOnConnectionLost(fileOperation)
-}
-
-func (r *retryableFileManagerImpl) Reset() error {
-	return r.fileManager.Reset()
+	return fm, nil
 }
 
-func (r *retryableFileManagerImpl) retryOnConnectionLost(fileOperation func() error) error {
+func (fm *fileManagerImpl) retryOnLostConnection(fileOperation func() error) error {
 	err := fileOperation()
 	if err == nil || !isConnectionLostError(err) {
 		return err // Operation successful or non-retryable error
 	}
 
-	if err := r.Reset(); err != nil {
+	if err := fm.Reset(); err != nil {
 		return err
 	}
 
@@ -159,14 +152,6 @@ func (r *retryableFileManagerImpl) retryOnConnectionLost(fileOperation func() er
 	return fileOperation()
 }
 
-func NewRetryableFileManager(config *SSHConfig) (FileManager, error) {
-	baseFileManager, err := NewFileManager(config)
-	if err != nil {
-		return nil, err
-	}
-	return &retryableFileManagerImpl{fileManager: baseFileManager}, nil
-}
-
 func isConnectionLostError(err error) bool {
 	return errors.Is(err, sftp.ErrSshFxConnectionLost)
 }
```

Reply:
This will require more code in `test.go`. Previously, we mocked the entire FileManager and simply tested the retry functionality. However, with this approach, we will need to mock at the client level (similar to how we currently test basic upload, download, and delete) rather than mocking the whole FileManager.

Reply:
Agreed on the code duplication part. We can remove the `retryFileManager` and drive the retry method based on a flag.
cc- @koladilip

Reply:
I am ok with this approach also

Reply:
addressed